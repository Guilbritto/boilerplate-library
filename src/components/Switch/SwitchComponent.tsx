import React, { useState } from 'react'

import { Container, CircleAura, Circle } from './SwitchComponent.styles'
import { SwitchProps } from './SwitchComponent.types'
import { useTheme } from '../../hooks/useTheme'
const SwitchComponent = ({
    isChecked,
    setIsChecked,
    disabled
}: SwitchProps) => {
    const [mouseDown, setMouseDown] = useState(false)
    const [mouseUp, setMouseUp] = useState(false)
    const theme = useTheme()

    const handleMouseUp = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        setMouseDown(false)
        setMouseUp(true)
    }

    const handleMouseDown = (
        e: React.MouseEvent<HTMLDivElement, MouseEvent>
    ) => {
        setMouseDown(true)
        setMouseUp(false)
    }

    const handleOnClick = () => {
        setIsChecked(!isChecked)
        setMouseDown(false)
        setMouseUp(false)
    }

    return (
        <Container
            isActive={isChecked}
            onClick={handleOnClick}
            disabled={disabled || false}
            theme={theme}
        >
            <CircleAura
                isActive={isChecked}
                disabled={disabled || false}
                isMouseDown={mouseDown}
                isMouseUp={mouseUp}
                onMouseDown={handleMouseDown}
                onMouseUp={handleMouseUp}
                theme={theme}
            >
                <Circle
                    theme={theme}
                    isActive={isChecked}
                    disabled={disabled || false}
                />
            </CircleAura>
        </Container>
    )
}

export default SwitchComponent
