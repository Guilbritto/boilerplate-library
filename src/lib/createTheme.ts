import { defaultTheme } from "../styles/defaultTheme";

export const createTheme = (theme: typeof defaultTheme) => {
  return {
    ...defaultTheme,
    colors: {
      ...defaultTheme.colors,
      ...theme.colors,
    },
    fontSizes: {
      ...defaultTheme.fontSizes,
      ...theme.fontSizes,
    },
  } as typeof defaultTheme;
}