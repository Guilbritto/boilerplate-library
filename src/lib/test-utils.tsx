import { render, RenderOptions } from '@testing-library/react'
import React, { ReactElement, FC } from 'react'
import { ThemeProvider } from 'styled-components'
import { defaultTheme } from '../styles/defaultTheme'

const Providers: FC = ({ children }) => {
    return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
}

/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: Providers, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
