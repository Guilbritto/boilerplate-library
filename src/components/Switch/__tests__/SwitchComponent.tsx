import { render, screen } from '../../../lib/test-utils'

import SwitchComponent from '../SwitchComponent'

describe('<SwitchComponent />', () => {
    it('Should be able to render a not active switch', () => {
        const { container } = render(<SwitchComponent isChecked={false} />)

        expect(container.firstChild).toMatchSnapshot()
    })
    it('Should be able to render a active switch', () => {
        const { container } = render(<SwitchComponent isChecked={true} />)

        expect(container.firstChild).toMatchSnapshot()
    })

    it('Should be able to render disabled switch', () => {
        const { container } = render(
            <SwitchComponent isChecked={false} disabled />
        )

        expect(container.firstChild).toMatchSnapshot()
    })
})
