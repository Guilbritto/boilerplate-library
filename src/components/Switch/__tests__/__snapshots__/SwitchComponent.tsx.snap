// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SwitchComponent /> Should be able to render a active switch 1`] = `
<div
  class="sc-bdfBQB rvVuK"
>
  <div
    class="sc-gsTEea kFvFwm"
  >
    <div
      class="sc-dlfnuX kFkJSD"
    />
  </div>
</div>
`;

exports[`<SwitchComponent /> Should be able to render a not active switch 1`] = `
<div
  class="sc-bdfBQB bZRwow"
>
  <div
    class="sc-gsTEea dNaYiJ"
  >
    <div
      class="sc-dlfnuX jzfzSw"
    />
  </div>
</div>
`;

exports[`<SwitchComponent /> Should be able to render disabled switch 1`] = `
<div
  class="sc-bdfBQB gaPvPz"
  disabled=""
>
  <div
    class="sc-gsTEea iVovoG"
    disabled=""
  >
    <div
      class="sc-dlfnuX djftuf"
      disabled=""
    />
  </div>
</div>
`;
